File: Top-Interview-Questions/LinkedList-Medium-Collection/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 * Runtime: 112 ms, faster than 90.80% of JavaScript online submissions for Add Two Numbers.
 * Memory Usage: 41.9 MB, less than 52.16% of JavaScript online submissions for Add Two Numbers.
 * TC: O(max(M, N)) where M is length of l1 and N is length of l2
 * SC: O(max(M, N)) where M is length of l1 and N is length of l2
 */
var addTwoNumbers = function(l1, l2, carry = 0) {
    if (!l1 && !l2) {
        if (carry > 0) {
            return new ListNode(carry, null);
        } else {
            return null;
        }
    }

    const l1Val = l1 ? l1.val : 0;
    const l2Val = l2 ? l2.val : 0;

    const sum = l1Val + l2Val + carry;
    const newNode = new ListNode(sum % 10, null);

    newNode.next = addTwoNumbers(l1 ? l1.next : null, l2 ? l2.next : null, sum >= 10 ? 1 : 0);

    return newNode;
};

File: Top-Interview-Questions/LinkedList-Medium-Collection/2-Add-Two-Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// TC: O(bigger size of listnode)
// SC: O(number of digits of sum)
// Runtime: 32 ms, faster than 75.70%
// Memory Usage: 70.3 MB, less than 22.08%
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        int carry = 0;
        
        ListNode *head, *curr;
        head = curr = NULL;
        
        while(l1 || l2){
            int sum = 0;
            
            if(l1){
                sum += l1->val;
                l1 = l1->next;
            }
            if(l2){
                sum += l2->val;
                l2 = l2->next;
            }
            
            sum += carry;
            
            if(sum > 9){
                sum = sum%10;
                carry = 1;
            }else{
                carry = 0;
            }   
            
            ListNode *newNode = new ListNode(sum);
            
            if(curr != NULL){
                curr->next = newNode;
                curr = curr->next;
            }
            else
                curr = newNode;
                
            if(head == NULL)
                head = curr;
        }
        
        if(carry){
            curr->next = new ListNode(1);
        }
            
        return head; 
    }
};